src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/Http/FrameOfT.cs

```diff
@@ -40,26 +47,47 @@ public Frame(IHttpApplication<TContext> application, FrameContext frameContext)
 
                     while (!_requestProcessingStopping)
                     {
-                        var result = await Input.ReadAsync();
-                        var examined = result.Buffer.End;
-                        var consumed = result.Buffer.End;
+                        if (needBuffer)
+                        {
+                            result = await Input.ReadAsync();
+                            buffer = result.Buffer;
+                            needBuffer = false;
+                        }
+
+                        var needAdvance = false;
 
                         try
                         {
-                            ParseRequest(result.Buffer, out consumed, out examined);
+                            ParseRequest(ref buffer, out consumed, out examined);
+
+                            if (buffer.Length == 0)
```

Comment:
Ah, I remember why I removed it.

`Frame` does not move the cursors anymore, it's the parser that does that. In the parser tests we have checks for incomplete request lines and headers and we assert the examined cursor there [[1](https://github.com/aspnet/KestrelHttpServer/blob/dd02ba577fdb17ff540863df87286062a33579b7/test/Microsoft.AspNetCore.Server.KestrelTests/HttpParserTests.cs#L62)] [[2](https://github.com/aspnet/KestrelHttpServer/blob/dd02ba577fdb17ff540863df87286062a33579b7/test/Microsoft.AspNetCore.Server.KestrelTests/HttpParserTests.cs#L200)]. So that's covered.

There's no issue with the parser either. It will always:

1. Either use the entire buffer and set `examined` to `buffer.End`; or
1. Slice the buffer and
   1. Either successfully parse the request line and set `examined` to the end of the slice, which is fine because it means that whatever awaits next will just complete immediately and get the remaining data in the buffer; or
   1. Reject the request because the entire request line or headers didn't fit in the slice, meaning they exceed their limits. At this point we don't care about advancing the reader properly anymore, since we're closing the connection.